prototype-to-fast-c++.rst

```diff
@@ -0,0 +1,292 @@
+Note that our process function expects a 1D input array. If you open a stereo .wav file, the array returned by wavfile.read will be 2D. In case this happens I'm only keeping the left channel. You can copy paste the following code into a main.py function:
+
+::
+
+    import time_scaling
+    import numpy as np
+    from scipy.io import wavfile
+
+    sRate, data = wavfile.read(r'/Users/jlaroche/temp/MessageInABottleMono.wav')
+    x=data[:,0] if data.ndim == 2 else data
+    factor = 1.2
+    out = time_scaling.process(x.astype(float)/32767,float(sRate),factor)
+    wavfile.write('./out.wav',sRate,(32767*out).astype(np.int16))
+
+You should be able to open the output file and listen to it in any program that plays wavfiles (for example afplay on macos).
+
+Let's time the function in ipython. For this you start ipython (install it if you don't have it, it's a great complement to python).
+In ipython, you can simply put %timeit in front of the line you'd like to benchmark:
+
+::
+
+    %timeit out = time_scaling.process(x.astype(float)/32767,float(sRate),factor)
+
+For the (quite long) wav file I was using, %timeit returned
+::
+
+    1 loop, best of 3: 15.1 s per loop
+
+
+Using pythran:
+__________________________________
+
+
+To be able to use the process function from the module that pythran will create, we need to export it to python. This is what the following #pythran export directive does. This can be placed anywhere in the .py file.
+
+::
+
+    #pythran export process(float[] or float[::],float,float)
+
+Note that the first parameter is declared as float[] or float[::] a simple float numpy array or a view into a float numpy array. The two remaining parameters are declared as float and it will be crucial to pass them as floats when calling process.
+
+Now simply run
+
+::
+
+    pythran time_scaling.py.
+
+A time_scaling.so file is created.
+Now the same main.py code will execute much faster because import time_scaling will now import a compiled, very efficient .so file.
+
+For the same file as above, %timeit now returns:
+::
+
+    1 loop, best of 3: 1.87 s per loop
+
+The speed up is amazing. The function runs about 14 times faster than it did in pure python/numpy.
+Note that if you pass an int instead of a float to the process function time_scaling.process(x.astype(float)/32767,int(sRate),factor) you will get a run-time error so make sure you're passing the very same types you've declared in time_scaling.py.
+
+
+Calling from C++
+__________________________________
+
+As I explained above, one of the most amazing aspects of pythran is that it generates self contained C++ code that can be called from any other C++ program. As I explained, the code is self contained in that it does not require any dlls, and makes no call to the python library. In short, it's a very efficient C++ version of your python/numpy algorithm, fully portable to any target architecture.
+To create a c++ version of our process function, we simply do:

```

Comment:
side note: it still requires headers that comes with pythran though (in the sense that you cannot ship the C++ file alone)